effectful-core/src/Effectful/Internal/Monad.hs

```diff
@@ -291,17 +307,34 @@ instance IOE :> es => MonadBaseControl IO (Eff es) where
 -- Primitive
 
 -- | Provide the ability to perform primitive state-transformer actions.
-data Prim :: Effect where
-  Prim :: Prim m r
+data Prim :: Effect
+
+type instance DispatchOf Prim = 'Static
+data instance DataR Prim = Prim
 
 -- | Run an 'Eff' computation with primitive state-transformer actions.
 runPrim :: IOE :> es => Eff (Prim : es) a -> Eff es a
-runPrim = evalData (DataA Prim)
+runPrim = evalData Prim
 
 instance Prim :> es => PrimMonad (Eff es) where
   type PrimState (Eff es) = RealWorld
   primitive = unsafeEff_ . IO
 
+----------------------------------------
+-- Dispatch
+
+-- | A dispatch type. For more information consult the documentation in
+-- "Effectful.Dispatch.Dynamic" and "Effectful.Dispatch.Static".
+data Dispatch = Dynamic | Static
+
+-- | A dispatch type of a particular effect.
+type family DispatchOf (e :: Effect) :: Dispatch
+
+-- | Internal representations of effects.
+type family EffectR (d :: Dispatch) = (r :: Effect -> Type) | r -> d where
+  EffectR 'Dynamic = HandlerR
+  EffectR 'Static  = DataR
+
```

Comment:
I really feel like there should be a 

```haskell
type Foo e = EffectR (DispatrchOf e) e
```

Reply:
I experimented with that, but I don't think it's worth it. It'd be used in 5 functions, the alias `StaticRepOf` doesn't save that much space and I find type synonyms can be confusing, since GHC expands them somewhat unexpectedly, e.g. in `ghci` `:t` expands type synonyms while `:i` doesn't.